src/MonteCarloMoves/random_walks.jl

```diff
@@ -0,0 +1,215 @@
+        end
+    end
+    return accept_this_walker, n_accept/n_steps, lattice
+end
+
+"""
+    MC_new_sample!(lattice::LatticeWalker, h::ClassicalHamiltonian, emax::Float64; energy_perturb::Float64=0.0)
+
+Generate a new sample for the lattice system.
+
+# Arguments
+- `lattice::LatticeWalker`: The walker to generate a new sample for.
+- `h::ClassicalHamiltonian`: The Hamiltonian containing the on-site and nearest-neighbor interaction energies.
+- `emax::Float64`: The maximum energy allowed for accepting a move.
+- `energy_perturb::Float64=0.0`: The energy perturbation used to make degenerate configurations distinguishable.
+
+# Returns
+- `accept_this_walker::Bool`: Whether the walker is accepted or not.
+- `lattice::LatticeWalker`: The updated walker.
+
+"""
+function MC_new_sample!(lattice::LatticeWalker{C},
+                        h::ClassicalHamiltonian,
+                        emax::Float64;
+                        energy_perturb::Float64=0.0,
+                        ) where C
+
+    accept_this_walker = false
+    emax = emax * unit(lattice.energy)
+
+    current_lattice = lattice.configuration
+    proposed_lattice = deepcopy(current_lattice)
+    generate_random_new_lattice_sample!(proposed_lattice)
+
+    perturbation_energy = energy_perturb * (rand() - 0.5) * unit(lattice.energy)
+    proposed_energy = interacting_energy(proposed_lattice, h) + perturbation_energy
+
+    @debug "proposed_energy = $proposed_energy, perturbed_energy = $(perturbation_energy), emax = $(emax)), accept = $(proposed_energy < emax)"
+
+    if proposed_energy < emax
+        lattice.configuration = proposed_lattice
+        lattice.energy = proposed_energy
+        accept_this_walker = true
+    end
+
+    return accept_this_walker, lattice
+end
+
+
+"""
+    generate_random_new_lattice_sample!(lattice::MLattice{C}) where C
+
+Generate a new random sample for the multi-component lattice system.
+
+# Arguments
+- `lattice::MLattice{C}`: The lattice system to generate a new sample for.
+
+# Returns
+- `lattice::MLattice{C}`: The updated lattice system.
+
+"""
+function generate_random_new_lattice_sample!(lattice::MLattice{C}) where C
+    occupancy = occupied_site_count(lattice)
+    # flush occupancy
+    for i in eachindex(lattice.components)
+        for j in eachindex(lattice.components[i])
+            lattice.components[i][j] = false
+        end
+    end
+    # generate a "look-up" table of unoccupied sites with elements get deleted as sites are occupied
+    # faster than shuffling a list of components
+    unoccupied = collect(1:num_sites(lattice))
+    for i in eachindex(lattice.components)
+        samples = sample(unoccupied, occupancy[i], replace=false, ordered=true)
+        for j in samples
+            lattice.components[i][j] = true
+            deleteat!(unoccupied, findfirst(isequal(j), unoccupied))
+        end
+    end
+    return lattice
+end
+
+function generate_random_new_lattice_sample!(lattice::SLattice)
```

Comment:
@yangmr04, many functions are missing docstrings. Does multiple dispatch automatically take care of this, or should we write some docstrings for these, too?

Reply:
I see. This is an **internal** function that should not be exposed to users. Thus, I don't police myself writing docstings for it. Not a good practice. All functions should have docstings. I will add them.